src/libsyntax/ast.rs

```diff
@@ -241,12 +241,13 @@ pub const CRATE_NODE_ID: NodeId = NodeId::from_u32_const(0);
 /// small, positive ids.
 pub const DUMMY_NODE_ID: NodeId = NodeId::MAX;
 
-/// A modifier on a bound, currently this is only used for `?Sized`, where the
-/// modifier is `Maybe`. Negative bounds should also be handled here.
+/// A modifier on a bound, currently this is used for `?Sized`, where the
+/// modifier is `Maybe`. Negative bounds are also parsed for error reporting.
 #[derive(Copy, Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Debug)]
 pub enum TraitBoundModifier {
     None,
     Maybe,
+    ErroneousNegative,
```

Comment:
Could you avoid adding impossible state to AST and report the error immediately in the parser, rather than in AST validation?

Reply:
I started out that way, but a parse error aborted parsing right away, thus I had to split the tests into 2 files. I'll try going back to that and see how rustfix applies.